### ISSUE TITLE
Enable truly multithread operations on read

### ISSUE THREAD

**Issue by williamfgc** [ornladios/ADIOS2#2202]
**Why is this feature important?**
adios motivation has always been write performance on HPC simulations based on MPI. While there is little gain from multithreading write operations in this context (serialization + HPC systems policy to disable MPI operations from threads), there is opportunity to multithread the read parts. Essentially, many deserialization tasks that build on top of low-level byte acquisition functionality (e.g. POSIX thread-safe `read`) functions (memory clipping, and/or per variable group stepping) can run as independent threads. 

The goal is not to build a thread-safe API (with mutexes and barriers), but a truly multithreaded API exploiting parallel portions on the read side.

Initial thoughts: C++/Fortran API, C APIs would have an extra `_t` postfix. The following functions could take an extra threadID parameter:
- `Variable::SetSelection` 
- `Engine::BeginStep/EndStep/Get/CurrentStep`

`Engine::Open/Close` are by definition "global" and the idea is to reuse the same object/handlers safely.

```
engine.BeginStep(threadID);
var.SetSelection(start(threadID), count(threadID, threadID);
engine.Get(var, data(threadID), threadID);
engine.EndStep(threadID);
```

**What is the potential impact of this feature in the community?**
A truly multithreaded API would run tasks in parallel. Currently, adios2 doesn't allow thread-safe operations.

**Is your feature request related to a problem? Please describe.**
Yes, reading portions of the dataset in parallel is done for the event processing at SNS/HFIR workflows.

**Describe the solution you'd like and potential required effort**
First, discuss a common API without breaking the current API as it covers well the majority of current production cases that require performance. I'd be happy to provide the underlying implementation and investigate with current workflows of interest.

**Additional context**
Ideally the added API functions could be reused on the write side.

**Comment by germasch** [comment]
> adios motivation has always been write performance on HPC simulations based on MPI. While there is little gain from multithreading write operations in this context (serialization + HPC systems policy to disable MPI operations from threads)

It's kinda off-topic, but I wouldn't write off MPI_THREAD_MULTIPLE (it's supported by spectrum mpi, e.g.), and I think some version of fat node /  hybrid threaded + MPI is going to be with us for a while.

> The goal is not to build a thread-safe API (with mutexes and barriers), but a truly multithreaded API exploiting parallel portions on the read side.

I'd say it's kind of an odd mix if your library is partially usable from threads, and partially bad things are going to happen.

So I have two thoughts on that:

* On the SetSelection and then subsequent Get (or Put): Since these are typically used in a paired fashion, where one uses two calls to really describe one action ("Get this selection"), even without the multi-threading aspect I think it'd be nice to have a single API call that will do the job. 

In fact, in one of the WDMapp codes, I actually wrote a wrapper to simplify get/put for me:
```fortran
  subroutine adios2_file_t_get_selection_n(this, varname, data, start, count)
    class(adios2_file_t) :: this
    character(*) :: varname
    complex :: data(*)
    integer, dimension(:) :: start, count

    type(adios2_variable_t) :: var
    integer :: ierr
    
    call this%inquire_variable(var, varname)
    call var%set_selection(start, count)
    call adios2_put(this%engine, var%var, data(1:1), ierr)
   [...]
```
It'd be easy to introduce a new function / new overload of Get that takes the selection, which wouldn't break compatibility, and that's then a function which can be made safe to use in a threaded environment.

* On thread safety in particular, I think it'd be pretty straightforward to follow the standard process to make a library threadsafe: Start by introducing a global lock (say per adios2::ADIOS) that ensures that only one user thread can enter the library at a time (ie., all the externally visible API functions will need to take that lock). That should get one most of the way there, other than providing alternatives to inherently unsafe APIs like SetSelection.

  And then, once the library is safe, but doesn't really take advantage of concurrency one push the locking dow to finer-grained locks like per I/O or per engine, etc. until one eventually actually allows concurrency where it really helps fixing bottlenecks, as guided by benchmarks.

  If one is concerned about the overhead of locking when the library is used in a single-threaded environment (which isn't likely to be an issue), one could build separate single-threaded vs thread-safe libs.

**Comment by williamfgc** [comment]
Thanks @germasch , good points.

> It's kinda off-topic, but I wouldn't write off MPI_THREAD_MULTIPLE (it's supported by spectrum mpi, e.g.), and I think some version of fat node / hybrid threaded + MPI is going to be with us for a while.

I think it's very much on-topic. If this is supported on Summit and you'd be using it for your application then write should also be taken into account.

> I'd say it's kind of an odd mix if your library is partially usable from threads, and partially bad things are going to happen.

Just breaking down tasks, we can start with read for practical purposes.

> even without the multi-threading aspect I think it'd be nice to have a single API call that will do the job

Let's open a separate issue and try to keep this focused on multithreading.

> On thread safety in particular, I think it'd be pretty straightforward to follow the standard process to make a library threadsafe

Yes, that's expected once an API is in place. Thanks.

**Comment by germasch** [comment]
> > It's kinda off-topic, but I wouldn't write off MPI_THREAD_MULTIPLE (it's supported by spectrum mpi, e.g.), and I think some version of fat node / hybrid threaded + MPI is going to be with us for a while.
> 
> I think it's very much on-topic. If this is supported on Summit and you'd be using it for your application then write should also be taken into account.

I believe MPI_THREAD_MULTIPLE is supported on Summit (spectrum mpi supports it, as do openmpi and mpich AFAIK.) The apps I'm working on aren't using it, even though at least one of them (XGC) does run hybrid MPI + OpenMP, and could potentially use multi-threaded I/O. I'm not aware of any need to do that, though, at this point.

> > even without the multi-threading aspect I think it'd be nice to have a single API call that will do the job
> 
> Let's open a separate issue and try to keep this focused on multithreading.

My point is that if you introduce an API that passes the selection right in the Get/Put call, you get a thread-safe interface, which may also be more convenient to use for some users (it simplifies at least two of my codes). Let me be more explicit about what I'm thinking of:

```cxx
    template <class T>
    void Get(Variable<T> variable, T *data, const adios2::Dims& start, const adios2::Dims& count,
             const Mode launch = Mode::Deferred);
```

* So one nice thing about this is that this API will work in single-threaded and in multi-threaded code without change. 
* The user also doesn't have to figure out their thread id to use it. If an app uses OpenMP, as many science codes do, they could use `omp_get_thread_num()`, but as opposed to OpenMP pragmas, this function doesn't exist if you compile without OpenMP, so it'll require some conditional compilation to use.
* Storing the selections by thread id inside of adios2 would be hard to implement efficiently. Essentially, what one needs is thread-local storage for the selection. The above API will pass the selection on the stack, which is automatically thread-local. For the other API, adios2 would have to emulate its own thread-local storage, e.g., by using a `std::map`. This will incur additional allocations, and walking the map will walk memory that has been allocated by different threads, which given that most systems are NUMA-aware these days is going to access memory that wasn't locally allocated, ie., is less efficient to access.

Also, another point I realized when reading the original proposal again now: I think threaded versions of BeginStep/EndStep are potentially confusing. What does it mean if one step is done with their step, but other threads aren't yet? I think there's not straightforward answer here, but since often a lot of the real work happens in PerformGets() in particular, there is probably a point in having that run on multiple threads at once, so essentially something like a "collective" `PerformGetsThreadCollective()` that needs to be called from all threads. I don't really know what a good answer is here...

**Comment by williamfgc** [comment]
@germasch we are mixing two separate things.

```
variable.SetSelection(start, count)
engine.Get(variable, data1, start1, count1);
engine.Get(variable, data2);
```
what's the second `Get` going to populate?  Please, open a separate issue as this is orthogonal to multithreading and deserves a lot more discussion as you can see. Let's stay on topic, so we can hear from everyone. Can we do that?

**Comment by williamfgc** [comment]
> The user also doesn't have to figure out their thread id to use it.

There has to be a handshake between a particular thread and specific operations within the adios2 library. The assumption is that each thread wants a particular selection and would do operations independent from other threads.

> this function doesn't exist if you compile without OpenMP, so it'll require some conditional compilation to use.

yes, that shouldn't be an issue, I have written a lot of portable code in the past with `ifdef _OPENMP` `threadID=0` would always exist.

> This will incur additional allocations, and walking the map will walk memory that has been allocated by different threads,

Synchronization is unavoidable, the assumption is that any parallel portion (memcpy/std::copy) would see benefits from parallelization that would offset any synchronization cost. I'd think that `read` is the biggest bottleneck, but we'd need to measure.

> I think threaded versions of BeginStep/EndStep are potentially confusing. What does it mean if one step is done with their step, but other threads aren't yet?

That each thread advances as its own pace avoiding barriers and race conditions. It also separates thread-local vs global state for current step. 

> there is probably a point in having that run on multiple threads at once, so essentially something like a "collective" PerformGetsThreadCollective() that needs to be called from all threads. I don't really know what a good answer is here...

Would a barrier after `PerformGets()` be sufficient?
